File: apps/ios-companion/Sources/Shared/GlanceFormat.swift
import Foundation

/// Tiny, deliberately plain text renderers over the glance payload.
///
/// NOTE (issue #7): these strings are PLACEHOLDER plumbing. What a glance
/// surface should actually say — wording, urgency, thresholds, when to stay
/// silent — is healthcare-domain UX reserved for the domain expert
/// (docs/design/WATCH-NOTIFICATIONS.ko.md; grammar: docs/PLAN.md §8.5).
public enum GlanceFormat {
    public static func scoreText(_ score: Int?) -> String {
        score.map(String.init) ?? "--"
    }

    /// "Energy 58 · high"
    public static func energyLine(_ payload: GlancePayload) -> String {
        "Energy \(scoreText(payload.energy.score)) · \(payload.energy.confidence.rawValue)"
    }

    /// "14:00 Deep work block [high]" — times rendered in the *server's*
    /// user timezone so phone/watch clock drift never lies about the plan.
    public static func nextBlockLine(_ payload: GlancePayload) -> String? {
        payload.nextBlocks.first.map { blockLine($0, timezone: payload.timezone) }
    }

    /// Same rendering for any single block (systemLarge lists all ≤3 blocks).
    public static func blockLine(_ block: GlanceBlock, timezone: String) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm"
        formatter.timeZone = TimeZone(identifier: timezone) ?? .current
        let title = block.title ?? (block.source == .proposal ? "Proposed block" : "Untitled block")
        var line = "\(formatter.string(from: block.start)) \(title)"
        if let demand = block.energyDemand {
            line += " [\(demand.rawValue)]"
        }
        return line
    }

    /// "2 alerts · Stress 82 vs baseline 55" / "No recent alerts"
    public static func alertsLine(_ payload: GlancePayload) -> String {
        let count = payload.alerts.unresolvedCount
        guard count > 0 else { return "No recent alerts" }
        let noun = count == 1 ? "alert" : "alerts"
        if let top = payload.alerts.top, !top.summary.isEmpty {
            return "\(count) \(noun) · \(top.summary)"
        }
        return "\(count) \(noun)"
    }

    /// Single-line variant for accessoryInline: "HM 58 · 2!"
    public static func inlineLine(_ payload: GlancePayload) -> String {
        var line = "HM \(scoreText(payload.energy.score))"
        if payload.alerts.unresolvedCount > 0 {
            line += " · \(payload.alerts.unresolvedCount)!"
        }
        return line
    }
}

File: apps/windows-companion/src/HealthMes.Glance.Core/GlanceFormat.cs
using System.Globalization;

namespace HealthMes.Glance.Core;

/// <summary>
/// Tiny, deliberately plain text renderers over the glance payload —
/// the C# twin of GlanceFormat.swift (iOS) / BriefingDisplayState (Android).
///
/// NOTE (issues #7/#11): these strings are PLACEHOLDER plumbing. What a
/// glance surface should actually say — wording, urgency, thresholds, when
/// to stay silent — is healthcare-domain UX reserved for the domain expert
/// (docs/design/WATCH-NOTIFICATIONS.ko.md; grammar: docs/PLAN.md §8.5).
/// </summary>
public static class GlanceFormat
{
    public const string NoScore = "--";

    /// <summary>Lowercase wire vocabulary ("high"/"medium"/"low").</summary>
    public static string WireName(this GlanceConfidence confidence) =>
        confidence.ToString().ToLowerInvariant();

    /// <summary>Lowercase wire vocabulary ("low"/"med"/"high").</summary>
    public static string WireName(this GlanceEnergyDemand demand) =>
        demand.ToString().ToLowerInvariant();

    public static string ScoreText(int? score) =>
        score?.ToString(CultureInfo.InvariantCulture) ?? NoScore;

    /// <summary>"Energy 58 · high"</summary>
    public static string EnergyLine(GlancePayload payload) =>
        $"Energy {ScoreText(payload.Energy.Score)} · {payload.Energy.Confidence.WireName()}";

    /// <summary>
    /// "14:00 Deep work block [high]" — times rendered in the SERVER's user
    /// timezone so desktop clock drift never lies about the plan. Unknown
    /// IANA ids fall back to the machine's local zone (mirrors iOS).
    /// </summary>
    public static string BlockLine(GlanceBlock block, string timezone)
    {
        var zone = ResolveTimeZone(timezone);
        var local = TimeZoneInfo.ConvertTime(block.Start, zone);
        var title = block.Title
            ?? (block.Source == GlanceBlockSource.Proposal ? "Proposed block" : "Untitled block");
        var line = $"{local.ToString("HH:mm", CultureInfo.InvariantCulture)} {title}";
        if (block.EnergyDemand is { } demand)
        {
            line += $" [{demand.WireName()}]";
        }
        return line;
    }

    public static string? NextBlockLine(GlancePayload payload) =>
        payload.NextBlocks.Count > 0 ? BlockLine(payload.NextBlocks[0], payload.Timezone) : null;

    /// <summary>"2 alerts · Stress 82 vs baseline 55" / "No recent alerts"</summary>
    public static string AlertsLine(GlancePayload payload)
    {
        var count = payload.Alerts.UnresolvedCount;
        if (count <= 0)
        {
            return "No recent alerts";
        }
        var noun = count == 1 ? "alert" : "alerts";
        var top = payload.Alerts.Top;
        return top is not null && top.Summary.Length > 0
            ? $"{count} {noun} · {top.Summary}"
            : $"{count} {noun}";
    }

    /// <summary>Single-line variant for one-slot surfaces: "HM 58 · 2!"</summary>
    public static string InlineLine(GlancePayload payload)
    {
        var line = $"HM {ScoreText(payload.Energy.Score)}";
        if (payload.Alerts.UnresolvedCount > 0)
        {
            line += $" · {payload.Alerts.UnresolvedCount}!";
        }
        return line;
    }

    /// <summary>
    /// Tray tooltip: "HealthMes · Energy 58 · high · 2 alerts", clamped to
    /// the 127-char NotifyIcon.Text budget.
    /// </summary>
    public static string TrayTooltip(GlancePayload? payload)
    {
        if (payload is null)
        {
            return "HealthMes · not paired";
        }
        var count = payload.Alerts.UnresolvedCount;
        var alerts = count == 1 ? "1 alert" : $"{count} alerts";
        var text = $"HealthMes · {EnergyLine(payload)} · {alerts}";
        return text.Length <= 127 ? text : text[..127];
    }

    public static TimeZoneInfo ResolveTimeZone(string timezone)
    {
        try
        {
            // .NET 8 resolves IANA ids on every OS (ICU); "UTC" also works.
            return TimeZoneInfo.FindSystemTimeZoneById(timezone);
        }
        catch (Exception error) when (error is TimeZoneNotFoundException or InvalidTimeZoneException)
        {
            return TimeZoneInfo.Local;
        }
    }
}
